mlir/lib/Gradient/Transforms/ConversionPatterns.cpp

```diff
@@ -220,145 +166,301 @@ struct BackpropOpPattern : public OpConversionPattern<BackpropOp> {
 
-        // The results of backprop are in data in
-        rewriter.create<LLVM::CallOp>(loc, backpropFnDecl, callArgs);
-        rewriter.eraseOp(op);
-        return success();
+    /// This registers a custom allocation and deallocation functions with Enzyme. It creates a
+    /// global LLVM array that Enzyme will convert to the appropriate metadata using the
+    /// `preserve-nvvm` pass.
+    ///
+    /// This functionality is described at:
+    /// https://github.com/EnzymeAD/Enzyme/issues/930#issuecomment-1334502012
+    LLVM::GlobalOp insertEnzymeAllocationLike(OpBuilder &builder, ModuleOp moduleOp, Location loc,
+                                              StringRef allocFuncName, StringRef freeFuncName) const
+    {
+        MLIRContext *context = moduleOp.getContext();
+        Type indexType = getTypeConverter()->getIndexType();
+        OpBuilder::InsertionGuard insertGuard(builder);
+        builder.setInsertionPointToStart(moduleOp.getBody());
+
+        auto allocationLike = moduleOp.lookupSymbol<LLVM::GlobalOp>(enzyme_allocation_key);
+        if (allocationLike) {
+            return allocationLike;
+        }
+
+        auto ptrType = LLVM::LLVMPointerType::get(context);
+        auto resultType = LLVM::LLVMArrayType::get(ptrType, 4);
+
+        builder.create<LLVM::GlobalOp>(loc, LLVM::LLVMArrayType::get(builder.getI8Type(), 3), true,
+                                       LLVM::Linkage::Linkonce, "dealloc_indices",
+                                       builder.getStringAttr(StringRef("-1", 3)));
+        allocationLike =
+            builder.create<LLVM::GlobalOp>(loc, resultType,
+                                           /*isConstant=*/false, LLVM::Linkage::External,
+                                           enzyme_allocation_key, /*address space=*/nullptr);
+        builder.createBlock(&allocationLike.getInitializerRegion());
+        Value allocFn = builder.create<LLVM::AddressOfOp>(loc, ptrType, allocFuncName);
+        Value sizeArgIndex =
+            builder.create<LLVM::ConstantOp>(loc, indexType, builder.getIndexAttr(0));
+        Value sizeArgIndexPtr = builder.create<LLVM::IntToPtrOp>(loc, ptrType, sizeArgIndex);
+        Value deallocIndicesPtr =
+            builder.create<LLVM::AddressOfOp>(loc, ptrType, "dealloc_indices");
+        Value freeFn = builder.create<LLVM::AddressOfOp>(loc, ptrType, freeFuncName);
+
+        Value result = builder.create<LLVM::UndefOp>(loc, resultType);
+        result = builder.create<LLVM::InsertValueOp>(loc, result, allocFn, 0);
+        result = builder.create<LLVM::InsertValueOp>(loc, result, sizeArgIndexPtr, 1);
+        result = builder.create<LLVM::InsertValueOp>(loc, result, deallocIndicesPtr, 2);
+        result = builder.create<LLVM::InsertValueOp>(loc, result, freeFn, 3);
+        builder.create<LLVM::ReturnOp>(loc, result);
+
+        return allocationLike;
+    }
+
+    static void insertFunctionName(PatternRewriter &rewriter, Operation *op, StringRef key,
```

Comment:
This could use a docstring. Something like `defineGlobalString` or similar for the function name would also be clearer I think.